onnxruntime/test/providers/internal_testing/internal_testing_tests.cc

```diff
@@ -0,0 +1,238 @@
+  bool enable_custom_ep = true;
+
+  CreateSession(SessionOptions{}, session, ort_model_path, enable_custom_ep);
+  ExecuteMnist(*session, enable_custom_ep);
+}
+
+// test that is the custom EP cannot take all nodes due to device limitations
+// that we fallback to the CPU implementations and can execute the model
+TEST(InternalTestingEP, TestLoadOrtModelWithReducedOpCoverage) {
+  const ORTCHAR_T* ort_model_path = ORT_TSTR("testdata/mnist.ort");
+  const std::unordered_set<std::string> supported_ops{"Conv", "Add", "Relu" /*, "MaxPool"*/};
+
+  std::unique_ptr<InferenceSessionWrapper> session;
+  bool enable_custom_ep = true;
+
+  CreateSession(SessionOptions{}, session, ort_model_path, enable_custom_ep, &supported_ops);
+
+  const auto& graph = session->GetGraph();
+  // Conv+Add gets fused by level 1 optimizer into single node. The 'Conv'/'Add'/'Relu' nodes should be compiled and
+  // handled by the custom EP. fallback to CPU for MaxPool.
+  ASSERT_EQ(graph.NumberOfNodes(), 6);
+  const auto& func_mgr = session->GetSessionState().GetFuncMgr();
+  NodeComputeInfo* compute_func = nullptr;
+
+  for (const auto& node : graph.Nodes()) {
+    EXPECT_EQ(supported_ops.count(node.OpType()), size_t(0))
+        << "Nodes with supported op types should have been replaced. Node with type " << node.OpType() << " was not.";
+    if (node.GetExecutionProviderType() == kInternalTestingExecutionProvider) {
+      EXPECT_STATUS_OK(func_mgr.GetFuncs(node.Name(), compute_func));
+      EXPECT_NE(compute_func, nullptr);
+    }
+  }
+
+  ExecuteMnist(*session, enable_custom_ep);
+}
+
+TEST(InternalTestingEP, TestMinimalRegistrationOfEPwithGetCapability) {
+  // TODO: In a full build we want to be able to call GetCapability for the NNAPI EP and produce an ORT format model
+  // with nodes correctly preserved. That requires being able to do a minimal registration of that EP where
+  // GetCapability is fully implemented, but Compile is a stub that just throws NOT_IMPLEMENTED if someone attempts
+  // to execute a model in that InferenceSession.
+}
+
+// count nodes assigned to the test EP and make sure they all have valid compute funcs
+static int CountAndValidateAssignedNodes(const Graph& current_graph,
+                                         const std::unordered_set<std::string>& supported_ops,
+                                         const FuncManager& func_mgr) {
+  int count = 0;
+
+  for (const auto& node : current_graph.Nodes()) {
+    EXPECT_EQ(supported_ops.count(node.OpType()), size_t(0))
+        << "Nodes with supported op types should have been replaced. Node with type " << node.OpType() << " was not.";
+    if (node.GetExecutionProviderType() == kInternalTestingExecutionProvider) {
+      NodeComputeInfo* compute_func = nullptr;
+      EXPECT_STATUS_OK(func_mgr.GetFuncs(node.Name(), compute_func));
+      EXPECT_NE(compute_func, nullptr);
+      ++count;
+    }
+
+    if (node.ContainsSubgraph()) {
```

Comment:
>if [](start = 4, length = 2)

else if?

Reply:
In this case as currently implemented it can't be both, but they're conceptually separate tests so I don't want to couple them with an `else if`. The testing EP could change to test taking control flow nodes, at which point the `else if` would become invalid (which is a hint that we should keep them separate).

---
In reply to: [523839104](https://github.com/microsoft/onnxruntime/pull/5776#discussion_r523839104) [](ancestors = 523839104)